### ISSUE TITLE
Rotations not working as expected? (Tracking Overrides w/ Custom Tracking)

### ISSUE THREAD

**Issue by RaytracerDale** [ValveSoftware/openvr#1651]
Hi there,

I'm building a driver using tracking overrides with the index to implement a custom tracking system. I'm having trouble getting the rotations in the correct universe for what the final driver pose expects.

I've got some code that converts a rotation matrix to a quaternion and I'm using that same code in Unity to convert the rotation matrix to a quaternion and display it in the editor, see a example here:

**Comment by RaytracerDale** [comment]
hey thanks for the quick reply! You certainly make a great point haha, its super easy to mess up the quaternion math 😆.

At the moment I'm already using Rotation matrices right up until actually updating the pose.  Here's a quick breakdown of what you see in the video:

Step 1. For each transform change I'm generating a model/TRS matrix inside of unity using the `Matrix4x4.TRS` function using the orange capsule’s transform.

Step 2. Send that matrix over TCP to the driver which actually performs the RotationMatrix->Quaternion conversion right before the tracker sends the pose update - this results in that weird fixed axis phenomena.

Step -2.5 Back in unity (just after sending the matrix packets), I'm using the same algorithm/code (just translated to C#) to compute the quaternion and then setting the transform.rotation of the green capsule to that newly computed quat. 

Is there any reason why the math’s would work in Unity and produce the expected rotation, but inside of SteamVR it would result in that weird fixed axes thing?

**Comment by RaytracerDale** [comment]
And for context, I've tried a couple of implementations of this same operation. Here's a gist of the two main working ones.

[Quaternion Implementations i've tried](https://gist.github.com/RaytracerDale/9db5364942b1c3440873f058a14cedd1)

For context, both of these work exactly the same. But result in the same problem inside of SteamVR.

So pretty safe to say i'm 90% sure its not the maths and something more fundamental.

**Comment by okawo80085** [comment]
Yeah looking at your math it's good, the only other thing coming to mind is maybe the order of operations maybe, but that should not matter also if you don't do quaternion math on the driver side, unless you're using relative rotations in unity (which won't work, think of drivers always providing rotation in the world global frame)

**Comment by RaytracerDale** [comment]
Sweet, good to know the math is solid, and just for the sake of clarity when you say relative rotations; are you referring to like a parent-child relationship and for example using the child’s localRotation. (Forgive me if this is a dumb question)

Because all the objects in the scene in unity are root level and have no other rotations applied. Is there some weird/wrong with how I’m making a rotation matrix from the rotation in unity? just calling Matrix4x4.TRS with the position world space rotation and vector3.one for scale. The quaternion code only uses the top 3x3 so having the full 4x4 shouldn’t matter.

I’m assuming that world space = world global frame, is that correct or are they not quite the same, again forgive me if this is an obvious dumb question 😆

This is a very peculiar issue, that you so much for your help so far by the way!

**Comment by RaytracerDale** [comment]
I’m not actually sending the quaternion, I’m sending the rotation matrix and to be sure it’s getting properly interpreted on the others side I have it serialising to json at the moment, it’s definitely making its way to SteamVR correctly, the idea is to simulate the same environment as what the tracker would send from inside Unity so Unity just hijacks the same tcp server as the tracker would normally use to send a tracking message.

Right now all I’m doing is taking that matrix packet deserialising it and running the exact same code on as on the unity side which is that QuaternionFromMatrix function and then piping the 
result of that into a quaternion like this: `vr::HmdQuaternion_t poseQuat = {quat.w, quat.x, quat.y, quat.z};`

is there anything else that could cause this behaviour?

**Comment by okawo80085** [comment]
Ok now im concerned about the math again... but also your pose struct, can you show your entire pose struct that gets used as the device's pose?

And about what else could be causing this, the nature of quaternions mostly, you see it's hard to notice if you messed up in quaternion transformations because unless you go through your entire range of motion you might miss it.

Gotta warn ya though, even if turns out all of my previous concerns are null, there is a chance you're getting bad matrices from unity, which are fine for unity, but don't work well with OpenVR, that can be checked too though.
Just replace your unity provided matrices with matrices that you define by hand and check if the orientation of the quaternions aligns with what you set in your matrix, if not *boom* thats your answer, your conversion math is messing up somewhere.

**Comment by RaytracerDale** [comment]
hi, hope you had a good easter!

Here's my full driver pose:
```
vr::DriverPose_t outPose = { 0 };

outPose.deviceIsConnected = true;
outPose.poseIsValid = true;
outPose.result = vr::ETrackingResult::TrackingResult_Running_OK;
outPose.willDriftInYaw = false;
outPose.shouldApplyHeadModel = false;
outPose.qDriverFromHeadRotation = { 1.0, 0.0, 0.0, 0.0 };
outPose.qWorldFromDriverRotation = { 1.0, 0.0, 0.0, 0.0 };

vr::HmdQuaternion_t quat = QuaternionFromMatrix(rotationMatrix);
outPose.qRotation = quat;

outPose.vecPosition[0] = 0;
outPose.vecPosition[1] = 1.89f;
outPose.vecPosition[2] = 0;

```

What orientations do you think I should manually try? And is there any specific way for how to generate these matrices, the way I know is this way:
![image](https://user-images.githubusercontent.com/92415888/164643892-266f9eab-d6e8-4d91-971d-e0313acccdd9.png)

And is there any special format to be weary of with the quaternions or matrices?

Here's another video with a bit more info as well:
https://www.youtube.com/watch?v=AKa48tk4JtM

**Comment by okawo80085** [comment]
Sorry for the late reply, your pose struct looks good

> What orientations do you think I should manually try? And is there any specific way for how to generate these matrices, the way I know is this way

You should start with every axis individually first, then start combining them with matrix multiplication

> And is there any special format to be weary of with the quaternions or matrices?

There was a really good issue in pyrr's repo about that, i'll link it if i find it again.

> Here's another video with a bit more info as well

Yeah so inverted axis is normal and it looks like your unity coordinate system is rotated 180 degrees around the y axis, that'd cause pitch and yaw to be inverted at the same time, apply a 180 rotation around the y axis to your matrix using matrix multiplication before you convert it quaternion.

**Comment by Flutterish** [comment]
Hey, maybe this wont really help, but I had to flip both w and z to get the rotation to work correctly. You said you only flipped w.
